kubernetes/resource_kubernetes_secret_v1_data.go

```diff
@@ -0,0 +1,249 @@
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// getting the secret data
+	res, err := conn.CoreV1().Secrets(namespace).Get(ctx, name, v1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return diag.Diagnostics{{
+				Severity: diag.Warning,
+				Summary:  "Secret deleted",
+				Detail:   fmt.Sprintf("The underlying secret %q has been deleted. You should recreate the underlying secret, or remove it from your configuration.", name),
+			}}
+		}
+		return diag.FromErr(err)
+	}
+
+	configuredData := d.Get("data").(map[string]interface{})
+
+	// stripping out the data not managed by Terraform
+	fieldManagerName := d.Get("field_manager").(string)
+
+	managedSecretData, err := getManagedSecretData(res.GetManagedFields(), fieldManagerName)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	data := res.Data
+	for k := range data {
+		_, managed := managedSecretData["f:"+k]
+		_, configured := configuredData[k]
+		if !managed && !configured {
+			delete(data, k)
+		}
+
+	}
+	decodedData := make(map[string]string, len(data))
+	for k, v := range data {
+		decodedData[k] = string(v)
+	}
+
+	d.Set("data", decodedData)
+
+	return nil
+}
+
+// getManagedSecretData reads the field manager metadata to discover which fields we're managing
+func getManagedSecretData(managedFields []v1.ManagedFieldsEntry, manager string) (map[string]interface{}, error) {
+	var data map[string]interface{}
+	for _, m := range managedFields {
+		// Only consider entries managed by the specified manager
+		if m.Manager != manager {
+			continue
+		}
+		var mm map[string]interface{}
+		err := json.Unmarshal(m.FieldsV1.Raw, &mm)
+		if err != nil {
+			return nil, err
+		}
+		// Check if the "data" field exists and extract it
+		if l, ok := mm["f:data"].(map[string]interface{}); ok {
+			data = l
+		}
+	}
+	return data, nil
+}
+
+func resourceKubernetesSecretV1DataUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	conn, err := m.(KubeClientsets).MainClientset()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	name := metadata.GetName()
+	namespace := metadata.GetNamespace()
+
+	_, err = conn.CoreV1().Secrets(namespace).Get(ctx, name, v1.GetOptions{})
+	if err != nil {
+		if d.Id() == "" {
+			// If we are deleting then there is nothing to do if the resource is gone
+			return nil
+		}
+		if statusErr, ok := err.(*errors.StatusError); ok && errors.IsNotFound(statusErr) {
+			return diag.Errorf("The Secret %q does not exist", name)
+		}
+		return diag.Errorf("Have got the following error while validating the existence of the Secret %q: %v", name, err)
+	}
+
+	// Craft the patch to update the data
+	dataInterface := d.Get("data")
+	data, ok := dataInterface.(map[string]interface{})
+	if !ok {
+		return diag.Errorf("Error casting data to map[string]interface{}")
```

Comment:
I probably wouldn't put Go syntax in an error message. This error is going to be pretty meaningless to the end-user.

In a lot of place we don't actually check the type assertion on the attributes. It's predetermined by the schema what the type is going to be, so the chances of running into this error are slim other than because of schema changes, e.g. https://github.com/hashicorp/terraform-provider-kubernetes/blob/01f6171db64bf3b57f61f6942b6a910905fd1b8a/kubernetes/resource_kubernetes_config_map_v1_data.go#L105